### PR TITLE
Bug 1138815 - fine tune the swiping behavior of swipeable panel view.

### PR DESCRIPTION
**Pull request by RudyLu**

### CODE REVIEW

**Comment by try-server-hook** [comment]
[RudyLu Rudy Lu (RudyLu)](https://github.com/RudyLu) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=7b00acea9b65)

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2a4564376774ab098ad470c94250fb8ccc360086

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=4dc11af2e59b0def94d0f150eb414777af120c6c

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=c030d0c3f91621df5583f330c75915f9e2d03391

File: apps/keyboard/js/views/swipeable_panel_view.js [hunk 1]
```diff
@@ -28,12 +30,22 @@ SwipeablePanelView.prototype.KEY_COUNT_PER_PANEL = 18;
...
 SwipeablePanelView.prototype.render = function() {
   var panel = document.createElement('div');
   panel.classList.add('swipe-panel');
 
   panel.addEventListener('touchstart', this);
 
+  this.swipingDetector = new SwipingDetector(panel);
+  this.swipingDetector.start();
+
+  panel.addEventListener('pan', this);
```
Comment (timdream) [inline apps/keyboard/js/views/swipeable_panel_view.js]: We are not using DOMEvents in Keyboard app. Please use `this.swipingDetector.onpan`, and `this.swipingDetector.onswipe` instead.

File: apps/keyboard/js/views/swiping_detector.js [hunk 2]
```diff
@@ -0,0 +1,152 @@
+'use strict';
+
+(function(exports) {
+
+function SwipingDetector(element) {
+  this.element = element;
+  this.state = 'init';
```
Comment (timdream) [inline apps/keyboard/js/views/swiping_detector.js]: `this.state = this.STATE_INIT` and
`SwipingDetector.prototype.STATE_INIT = 0`

is the better approach for a enum value.

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=fa5de78bef3303fbabba4d5ee2fbd534ea92fe89

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=0fd306bc4eb412012c422599699736861b6fe413

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=38451b386644cfc9afc3938499606033300ab532

File: apps/keyboard/js/views/swipeable_panel_view.js [hunk 3]
```diff
@@ -100,7 +112,7 @@ SwipeablePanelView.prototype._moveSection = function(index, distance) {
   style.transform = 'translateX(' + distance + 'px)';
   style.transition = 'transform ' + 300 + 'ms ease';
 };
-
+
```
Comment (timdream) [inline apps/keyboard/js/views/swipeable_panel_view.js]: whitespace

File: apps/keyboard/js/views/swipeable_panel_view.js [hunk 4]
```diff
@@ -28,12 +30,22 @@ SwipeablePanelView.prototype.KEY_COUNT_PER_PANEL = 18;
...
 SwipeablePanelView.prototype.render = function() {
   var panel = document.createElement('div');
   panel.classList.add('swipe-panel');
 
   panel.addEventListener('touchstart', this);
 
+  this.swipingDetector = new SwipingDetector(panel);
+  this.swipingDetector.start();
+
```
Comment (timdream) [inline apps/keyboard/js/views/swipeable_panel_view.js]: space

File: apps/keyboard/js/views/swipeable_panel_view.js [hunk 5]
```diff
@@ -110,100 +122,93 @@ SwipeablePanelView.prototype._updateIndicator = function() {
 };
 
 SwipeablePanelView.prototype.handleEvent = function(evt) {
-  var totalWidth;
-
   switch (evt.type) {
     case 'touchstart':
-      this.touchStartTimestamp = evt.timeStamp;
       this.startX = evt.touches[0].pageX;
       this.deltaX = 0;
```
Comment (timdream) [inline apps/keyboard/js/views/swipeable_panel_view.js]: `touchstart` should be abstracted in `SwipingDetector`; `SwipeablePanelView` should not be listening and keeping these properties.

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b7023e448d4a1d3fc6a4dcd7429b5b19e6e303cb

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=862ea761fcd2c19f2f6adec84cea3d9a1f292000

File: apps/keyboard/js/views/swipeable_panel_view.js [hunk 6]
```diff
@@ -109,101 +119,90 @@ SwipeablePanelView.prototype._updateIndicator = function() {
...
-        } else if (!forward && this.currentSectionIndex > 0) {
-          targetIndex = this.currentSectionIndex - 1;
-        }
-      }
+  var targetIndex = this.currentSectionIndex;
+  var forward = evt.direction === 'left';
 
-      this.gotoSection(targetIndex);
-      break;
+  var farenough = Math.abs(this.deltaX) > totalWidth * this.SWIPE_FRACTION;
```
Comment (timdream) [inline apps/keyboard/js/views/swipeable_panel_view.js]: `isFarEnough`?

File: apps/keyboard/js/views/swipeable_panel_view.js [hunk 7]
```diff
@@ -109,101 +119,90 @@ SwipeablePanelView.prototype._updateIndicator = function() {
...
+  var velocity = evt.vx;
+  var fastenough = Math.abs(velocity) > this.SWIPE_SPEED;
+  var samedirection = velocity === 0 || this.deltaX / velocity >= 0;
```
Comment (timdream) [inline apps/keyboard/js/views/swipeable_panel_view.js]: `isSameDirection`

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b459de93dd674bad445eee86559c9a5ba48c63cb

**Comment by autolander** [comment]
http://docs.taskcluster.net/tools/task-graph-inspector/#HT0yyqwfSaCTTq0u4wlhrA

The pull request failed to pass integration tests. It could not be landed, please try again.

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=91f4fcd12d2377f40ca50b9dedba4a4d886f9e47

**Pull request merged**